Source: Gentoo
Upstream: no
Reason: Patch Gentoo nicked from Debian recode-3.6-11. Needed for fortune-mod.
--- src/libiconv.c
+++ src/libiconv.c
@@ -195,12 +195,17 @@
 	 memcpy() doesn't do here, because the regions might overlap.
 	 memmove() isn't worth it, because we rarely have to move more
 	 than 12 bytes.  */
-      if (input > input_buffer && input_left > 0)
+      cursor = input_buffer;
+      if (input_left > 0)
 	{
-	  cursor = input_buffer;
-	  do
-	    *cursor++ = *input++;
-	  while (--input_left > 0);
+	  if (input > input_buffer)
+	    {
+	      do
+		*cursor++ = *input++;
+	      while (--input_left > 0);
+	    }
+	  else
+	    cursor += input_left;
 	}
     }
 
--- src/request.c
+++ src/request.c
@@ -1073,7 +1073,7 @@
   if (task->output.cursor + 4 >= task->output.limit)
     {
       RECODE_OUTER outer = task->request->outer;
-      size_t old_size = task->output.limit - task->output.buffer;
+      size_t old_size = task->output.cursor - task->output.buffer;
       size_t new_size = task->output.cursor + 4 - task->output.buffer;
 
       /* FIXME: Rethink about how the error should be reported.  */
--- src/task.c
+++ src/task.c
@@ -1198,6 +1198,8 @@
       else
 	success = transform_mere_copy (subtask);
 
+      task->output = subtask->output;
+       
       if (subtask->input.name && *subtask->input.name)
 	fclose (subtask->input.file);
       if (subtask->output.name && *subtask->output.name)
--- src/hash.h
+++ src/hash.h
@@ -21,6 +21,11 @@
 /* Make sure USE_OBSTACK is defined to 1 if you want the allocator to use
    obstacks instead of malloc, and recompile `hash.c' with same setting.  */
 
+#define hash_lookup recode_hash_lookup
+#define hash_delete recode_hash_delete
+#define hash_free   recode_hash_free
+#define hash_insert recode_hash_insert
+
 #ifndef PARAMS
 # if PROTOTYPES || __STDC__
 #  define PARAMS(Args) Args
